### ISSUE TITLE
実験: 競りteamPointCards予測のlearning curveをデータ量別に評価する

### ISSUE THREAD

**Issue by eletim** [eletim/napoleon-web#377]
Related: #202, #373, #375
Base branch: `feature/issue-202-non-playing-ai`

## 背景

Issue #375 で、RuleBased / Conservative 50:50 の frozen bidding opponents を用い、最終役職 + candidate team の最終ポイント札枚数を教師にする role-decomposed model を実装・検証した。

初期実験は source states 1000 / raw samples 5552 / repeats=1 で、team point-card prediction は以下だった。

- MAE: 3.226
- RMSE: 4.079
- Pearson: 0.358
- same-target suit exact tie: 0%
- strongest-suit teacher-best: 47.8%

単一 terminal reward 教師より明確にsignalは改善したが、sample数はまだ小さい。

この段階では PPO のように policy update により次 iteration のstate分布を変える必要性はまだ薄い。一方で、学習データ量を段階的に増やし、同一validation条件で性能推移を観察する learning curve は取得したい。

## 目的

固定した teacher / opponent / model / validation 条件のもとで、training dataset の source state 数だけを段階的に増やし、teamPointCards予測・role予測・action rankingがどこまで改善するかを確認する。

これは policy-improvement iteration ではなく、**data scaling / learning curve experiment** とする。

## 基本方針

同一の大規模counterfactual dataset poolを一度生成し、その中からnested subsetを使って学習する。

例:

- 1k states
- 2k states
- 5k states
- 10k states
- 20k states

必要に応じて最大20k〜30k程度まで増やしてよい。

各subsetは必ず包含関係にする。

```text
train-1k ⊂ train-2k ⊂ train-5k ⊂ train-10k ⊂ train-20k
```

これにより、差がsample量によるものかを比較しやすくする。

## 重要: validation固定

learning curve評価ではvalidation setを毎回変えないこと。

推奨:

- 最初に固定held-out validation state群を確保
- すべてのdata-size runで同じvalidation stateKey集合を使用
- validation stateをどのtraining subsetにも混ぜない

可能なら1000〜2000 state程度の固定validationを確保する。

stateKey leakageがないことを必ず検証する。

## Dataset generation

Issue #375 schema v2 generatorを使用する。

Frozen bidding opponents:

- RuleBased 50%
- Conservative 50%
- per frozen seat independently sampled
- seeded deterministic

Downstream:

- playing: `ppo-separated-v1000`
- adjutant/exchange:現行frozen artifact

repeatsは #372 で downstream rollout noise がほぼ無いことが分かっているため、まず `1` とする。

Counterfactual action planは #375 と同じものを使う。

最低限coverage:

- PASS
- S/H/D/C
- target 13..19
- legal action 0..28

## Dataset size目標

基準案:

| checkpoint | training source states |
|---|---:|
| L1 | 1,000 |
| L2 | 2,000 |
| L3 | 5,000 |
| L4 | 10,000 |
| L5 | 20,000 |

1 stateあたり約5〜6 forced action rolloutなら、20k statesで約10万〜12万game rollout相当を想定する。

実行時間・storageが問題なければ20kまで行う。

## Model / training条件固定

sample数以外をできるだけ固定する。

Issue #375 primary MSE runを基準にする。

- model: `bidding-role-decomposed-q-mlp-v1`
- hidden dims: `[512,512,256,256]`
- value target: `team-point-cards`
- loss: MSE
- lr: `1e-3`
- batch size: `128`
- training device:利用可能なdevice

ただし、epoch数を各datasetで固定すると大datasetほどoptimizer stepが増えるため、比較方法を事前に整理すること。

最低限、以下2つのどちらを採用したか明示する。

### A. fixed epochs

同一epoch数で比較。

### B. fixed optimizer steps / sample exposures

総更新量を揃えて比較。

まずは practical learning curve として fixed epochs でもよいが、結果解釈でstep数差を明記する。

## Checkpointごとの評価

各training sizeで最低限以下を保存・報告する。

### TeamPointCards prediction

- MAE
- RMSE
- Pearson correlation
- role別 MAE/RMSE/Pearson
- suit別 MAE/RMSE/Pearson
- target別 MAE/RMSE/Pearson

### Role prediction

- accuracy
- macro F1
- cross entropy
- per-role precision/recall
- confusion matrix
- majority baseline
- action-index baseline

### Action-ranking diagnostics

teacher-covered actions内で:

- best-action hit
- top-3 hit
- pairwise ranking accuracy
- predicted best suit distribution
- predicted best target distribution
- strongest suit match

### Baselines

Issue #375同様:

- global mean
- role mean
- action-index mean
- role x action mean

を固定validationで比較する。

## Learning curve summary

最終的に以下のような表を作る。

| train states | raw samples | MAE | RMSE | Pearson | role acc | best hit | pairwise | strongest match |
|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| 1k | ... | ... | ... | ... | ... | ... | ... | ... |
| 2k | ... | ... | ... | ... | ... | ... | ... | ... |
| 5k | ... | ... | ... | ... | ... | ... | ... | ... |
| 10k | ... | ... | ... | ... | ... | ... | ... | ... |
| 20k | ... | ... | ... | ... | ... | ... | ... | ... |

グラフ化用のJSON/CSVを残してもよい。

### データ不足型

sampleを増やすほど、

- MAE/RMSEが継続改善
- Pearsonが継続上昇
- role prediction改善
- action ranking改善

する場合、現在のarchitecture/teacherを維持してさらにdata scalingする価値がある。

### 頭打ち型

5k〜10k以降ほぼ改善しない場合、sample追加よりも

- model architecture
- observation
- teacher decomposition
- action scoring

を見直す。

### 過学習型

train metricだけ改善しvalidationが悪化する場合、regularization/model capacity/epoch数を見直す。

## 重要: 今回はpolicy iterationではない

各checkpointの学習済みmodelを次のdataset生成policyとして使わない。

すべて同一のsource distribution / opponent topology / counterfactual planから生成する。

つまり、今回の「区切り」は PPO iteration ではなく **data-size checkpoint** である。

後続で必要なら、学習済みQ policyをdataset generationへ入れるactive/iterative learningを別Issueで扱う。

## 実行方法

可能なら大きいdataset poolを一度生成してnested subsetで再利用する。

これにより1k/2k/5k/10k/20kごとに同じゲームを再生成しない。

manifest/provenanceで:

- full pool identity
- fixed validation stateKey list
- 各training subset stateKey list/hash
- subset size
- source commit
- artifact SHA

を保存する。

## Tests / safety

恒久コード変更が不要なら実験のみでもよい。

learning-curve runner / subset toolingを追加する場合は最低限:

1. nested subset保証
2. validation leakageなし
3. deterministic subset selection
4. manifest/provenance
5. 同一validation使用確認
6. metric aggregation correctness

## Git / PR

まず既存CLIだけで実験可能か確認する。

- 恒久コード変更不要なら branch / PR不要、結果をIssueへ直接コメント
- reusable learning-curve runner等のコード追加が必要なら、最新 `feature/issue-202-non-playing-ai` からIssue専用branchを作成し、PR baseを同親branchにする
- mainにはmergeしない
- コード変更した場合はCodex review最大4回、修正最大3回

## 完了時に必ず報告

- full dataset pool state/sample数
- fixed validation state数
- training checkpoints
- 各checkpointのraw sample数
- 各checkpointのteamPointCards MAE/RMSE/Pearson
- role accuracy/macro F1
- best-action/top3/pairwise
- predicted suit/target distribution
- strongest suit match
- baseline比較
- learning curveが継続改善 / 頭打ち / 過学習のどれか
- 次にさらにsampleを増やすべきか
- それともarchitecture/action score設計へ進むべきか
- branch/PR/merge有無
- mainへmergeしていないこと
